### PR TITLE
First approach to add aggregation options to CartoDB layers

### PR DESCRIPTION
**Pull request by elenatorro**
## Project: https://github.com/CartoDB/product/issues/30

* Related issues: https://github.com/CartoDB/cartodb/issues/13683
* Related changes: [cartodb.js#13683-add-default-point-aggregation](https://github.com/CartoDB/cartodb.js/tree/13683-add-default-point-aggregation)

We have run some performance tests using chrome dev tools, to see how point aggregation improves rendering when having a high amount of points. We have run the test for 150k, 500k and 1M points. We have run six times every test and then we have calculated the average of the loading result in ms.

| Points | No Aggregation | Aggregation | Delta |
|--------|----------------|-------------|-------------|
|150000	 |13274.16667ms   |10971.66667ms|2303ms|
|500000	 |15762.33333ms   |13315.16667ms|2447ms|
|1000000 |19152.5ms       |16464.5ms    |2688ms|

We can see there is an improvement when using aggregation.

chart

## Discussion 

Currently, Windshaft triggers an error when the layer's geometry is not "point" --> ["Aggregation is available only for point geometries."](https://carto.com/docs/carto-engine/maps-api/tile-aggregation/#tile-aggregation). AFAIK, there is no way to know the geometry before sending the layer to Windshaft, so we have to discuss how to implement this cc @dgaubert @alonsogarciapablo @ivanmalagon

### CODE REVIEW

File: lib/assets/javascripts/deep-insights/api/create-dashboard.js [hunk 1]
```diff
@@ -134,6 +137,17 @@ var createDashboard = function (selector, vizJSON, opts, callback) {
   });
 };
 
+var _addAggregationToCartoDBLayers = function (vizJSON) {
+  var LAYER_TYPE = 'CartoDB';
+  var aggregation = new LayerAggregationModel();
+
+  vizJSON.layers.forEach(function (layer) {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/deep-insights/api/create-dashboard.js]: Changing the vizJSON via JS is not something we typically do... If we believe layers should have this info in the viz.json, let's add it to the viz.json generated by the backend.

Comment (elenatorro) [reply lib/assets/javascripts/deep-insights/api/create-dashboard.js]: That is one of the things I asked to @ivanmalagon, but the problem is that then all the vizJSON's will have this option by default. cc @javitonino

**Comment by alonsogarciapablo** [comment]
I was envisioning this as something we would do at the CARTO.js level (i.e: all maps rendered by CARTO.js would have point aggregation enabled by default - and benefit from faster point rendering). How does that sound?

**Comment by elenatorro** [comment]
In my opinion it sounds cleaner than changing the vizJSON directly in Builder. The solution would be to add this default options here -> https://github.com/CartoDB/cartodb.js/blob/940bfae1a3e1cf2f02b1c6ffb58fdad9af02cc66/src/geo/map/cartodb-layer.js#L45.

**Comment by alonsogarciapablo** [comment]
Sounds good to me!

About "Windshaft triggers an error when the layer's geometry is not "point""... should we open an issue in https://github.com/CartoDB/Windshaft-cartodb/?

Thanks!

**Comment by ivanmalagon** [comment]
Answering @alonsogarciapablo (even though @elenatorro already did, just to add more context)

1- Setting the aggregation at `carto.js` level would set the aggregation also for all CARTO.js custom applications. I was envisioning it as something that the client decides. In this case, Builder is the client and we're deciding that all maps will be aggregated. I don't know if it's a decision that we can take for the rest of application that will be made using CARTO.js v4.

2- About the vizJSON modification. Do you think that we will want to tweak the aggregations in the future via UI? I was thinking to prepare a path in this PR to connect Builder and CARTO.js to set aggregations. It's true that that will demand changes also in Backend to persist the aggregation and then no vizJSON modification will be needed, but this PR would pave the path.

3- If Windshaft don't return any error then there's no need right now to return aggregations in vizJSON by backend. Moreover, we don't have any persistence mechanism in backend for aggregations so backend would create a constant payload. That can be done in frontend.

Does that make sense?

**Comment by alonsogarciapablo** [comment]
> 1- Setting the aggregation at carto.js level would set the aggregation also for all CARTO.js custom applications. I was envisioning it as something that the client decides. In this case, Builder is the client and we're deciding that all maps will be aggregated. I don't know if it's a decision that we can take for the rest of apps that will be made using CARTO.js v4.

This is how I see it: default Point Aggregation is just a performance boost and it's transparent for users, that why I think it makes sense to "enable it" by default for all maps rendered by CARTO.js. But that's just my POV. If we decide to let the client decide, it means CARTO.js APIs (both the internal and external) will have to expose this somehow (eg: changes to viz.json).

> 2- About the vizJSON modification. Do you think that we will want to tweak the aggregations in the future via UI? I was thinking to prepare a path in this PR to connect Builder and CARTO.js to set aggregations. It's true that that will demand changes also in Backend to persist the aggregation and then no vizJSON modification will be needed, but this PR would pave the path.

That's a great question. We will probably want to allow users tweak aggregations from the UI to implement proper clustering, so the answer is YES. That is probably coming.

> 3- If Windshaft don't return any error then there's no need right now to return aggregations in vizJSON by backend. Moreover, we don't have any persistence mechanism in backend for aggregations so backend would create a constant payload. That can be done in frontend.

I wasn't suggesting we persist that. I was suggesting that the backend can generate this "default" aggregation if we follow this path. In the future, if we allow users to tweak the aggregations from the UI (see 2.) we'll somehow need to store this info somewhere (eg: backend).

**Comment by ivanmalagon** [comment]
Summoning @inigomedina 

We're arguing if the server aggregation should be done by default and transparently to CARTO.js applications. So if you don't do anything and your layer exceeds 100K rows, it'll be aggregated.

We're going that way in Builder but if CARTO.js does the same, our development changes.

What do you think? Do we enable aggregation by default transparently to CARTO.js apps?

**Comment by alonsogarciapablo** [comment]
>  if the server aggregation should be done by default and transparently to CARTO.js applications

Since this is something transparent for the final user., we can see it as a performance improvement that all CARTO.js apps can benefit from. Why would someone NOT want to benefit from this?

**Comment by inigomedina** [comment]
Yes, I follow the same argument exposed by @alonsogarciapablo. Maybe in the future we discover that we need to give some kind of flexibility, allowing to manage that `default`. But right now, it looks the natural way.

File: lib/assets/javascripts/builder/editor/layers/layer-aggregation-model.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  centroid: 'centroid',
+  grid: 'point-grid',
+  sample: 'point-sample'
+};
+
+module.exports = Backbone.Model.extend({
+  defaults: {
+    threshold: 100000,
+    resolution: 1,
+    placement: PLACEMENTS.sample
```
Comment (ivanmalagon) [inline lib/assets/javascripts/builder/editor/layers/layer-aggregation-model.js]: Default aggregation comes with no placement.

From the docs:
```
When no placement or columns are specified a special default aggregation is performed.
```

**Comment by elenatorro** [comment]
As @ivanmalagon mentions [here](https://github.com/CartoDB/cartodb/pull/13718#pullrequestreview-108582004) we do not need to pass a `placement` value to perform the aggregation. However, as it is said in the documentation, I have the following question:

> point-sample: This is the default placement. 

If we set `"point-sample"` as `placement`, Windshaft triggers an error if we try to change the aggregation style from Builder. But if we do not set any `placement` value, it works. If this is the default placement, why does it fail when we set it manually?

**Comment by ivanmalagon** [comment]
Just to have it here documented. As @dgaubert told us, metadata returned from TurboCarto it's affected by aggregations.

Here are two screenshots from a dataset of 100K rows.

### No aggregation

image

Speed goes from 0 to 254.

These numbers are right. We can check it with a simple SQL.

image

### Default aggregation

image

Speed goes now from 0 to 148. 😞

**Comment by elenatorro** [comment]
Yes, we have a problem here. While widgets seem not to be affected, legends show different information with and without aggregation.

* Without aggregation
captura de pantalla 2018-04-02 a las 14 43 56

* With aggregation

captura de pantalla 2018-04-02 a las 14 44 31

**Comment by dgaubert** [comment]
> If this is the default placement, why does it fail when we set it manually?

It looks like a bug. Don't you mind opening a ticket [here](https://github.com/CartoDB/Windshaft-cartodb)?

**Comment by alonsogarciapablo** [comment]
@elenatorro apart from the differences in the legends (and colors) it looks like there are actually more points when aggregation is enabled. How is that possible?

**Pull request closed without merging**